File: days/day_04/solutions/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func countNeighbors(grid []string, i, j, rows, cols int) int {
	count := 0
	for di := -1; di <= 1; di++ {
		for dj := -1; dj <= 1; dj++ {
			if di == 0 && dj == 0 {
				continue
			}
			ni, nj := i+di, j+dj
			if ni >= 0 && ni < rows && nj >= 0 && nj < cols && grid[ni][nj] == '@' {
				count++
			}
		}
	}
	return count
}

func solve(lines []string) (string, string) {
	rows := len(lines)
	cols := 0
	if rows > 0 {
		cols = len(lines[0])
	}

	// Part 1: Count accessible rolls (fewer than 4 neighbors that are '@')
	part1Count := 0
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if lines[i][j] == '@' {
				neighbors := countNeighbors(lines, i, j, rows, cols)
				if neighbors < 4 {
					part1Count++
				}
			}
		}
	}

	// Part 2: Iteratively remove accessible rolls until none can be removed
	grid := make([][]byte, rows)
	for i := range grid {
		grid[i] = make([]byte, cols)
		copy(grid[i], lines[i])
	}

	part2Count := 0
	for {
		toRemove := [][2]int{}
		for i := 0; i < rows; i++ {
			for j := 0; j < cols; j++ {
				if grid[i][j] == '@' {
					// Convert grid to string slice for neighbor counting
					gridStr := make([]string, rows)
					for k := range grid {
						gridStr[k] = string(grid[k])
					}
					neighbors := countNeighbors(gridStr, i, j, rows, cols)
					if neighbors < 4 {
						toRemove = append(toRemove, [2]int{i, j})
					}
				}
			}
		}

		if len(toRemove) == 0 {
			break
		}

		// Remove all marked positions
		for _, pos := range toRemove {
			grid[pos[0]][pos[1]] = '.'
		}
		part2Count += len(toRemove)
	}

	return strconv.Itoa(part1Count), strconv.Itoa(part2Count)
}

func main() {
	file, err := os.Open("../data/input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()
	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, strings.TrimSpace(scanner.Text()))
	}
	if err := scanner.Err(); err != nil {
		panic(err)
	}

	part1, part2 := solve(lines)
	fmt.Printf("Part 1: %s\n", part1)
	fmt.Printf("Part 2: %s\n", part2)
}

File: days/day_04/solutions/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;

sub read_input_raw {
    my ($file_path) = @_;
    open my $fh, '<', $file_path or die "Cannot open file: $!";
    my $content = do { local $/; <$fh> };
    close $fh;
    return $content;
}

sub count_neighbors {
    my ($grid, $i, $j, $rows, $cols) = @_;
    my $neighbor_count = 0;
    for my $di (-1..1) {
        for my $dj (-1..1) {
            next if $di == 0 && $dj == 0;
            my $ni = $i + $di;
            my $nj = $j + $dj;
            if ($ni >= 0 && $ni < $rows && $nj >= 0 && $nj < $cols) {
                if (substr($grid->[$ni], $nj, 1) eq '@') {
                    $neighbor_count++;
                }
            }
        }
    }
    return $neighbor_count;
}

sub solve {
    my ($input_data) = @_;
    my @lines = grep { $_ =~ /\S/ } split /\n/, $input_data;
    my $rows = scalar @lines;
    my $cols = $rows > 0 ? length($lines[0]) : 0;
    
    # Part 1: Count accessible rolls (fewer than 4 neighbors that are '@')
    my $part1_count = 0;
    for my $i (0..$rows-1) {
        for my $j (0..$cols-1) {
            if (substr($lines[$i], $j, 1) eq '@') {
                my $neighbors = count_neighbors(\@lines, $i, $j, $rows, $cols);
                if ($neighbors < 4) {
                    $part1_count++;
                }
            }
        }
    }
    
    # Part 2: Iteratively remove accessible rolls until none can be removed
    my @grid = map { [split //, $_] } @lines;
    my $part2_count = 0;
    
    while (1) {
        my @to_remove = ();
        my @grid_str = map { join('', @$_) } @grid;
        for my $i (0..$rows-1) {
            for my $j (0..$cols-1) {
                if ($grid[$i][$j] eq '@') {
                    my $neighbors = count_neighbors(\@grid_str, $i, $j, $rows, $cols);
                    if ($neighbors < 4) {
                        push @to_remove, [$i, $j];
                    }
                }
            }
        }
        
        last unless @to_remove;
        
        # Remove all marked positions
        for my $pos (@to_remove) {
            my ($i, $j) = @$pos;
            $grid[$i][$j] = '.';
        }
        $part2_count += scalar @to_remove;
    }
    
    return ($part1_count, $part2_count);
}

my $data = read_input_raw("../data/input.txt");
my ($part1, $part2) = solve($data);
print "Part 1: $part1\n";
print "Part 2: $part2\n";
